compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/RedundantGotoMethodTransformer.kt

```diff
@@ -22,31 +22,76 @@ import org.jetbrains.org.objectweb.asm.Opcodes
 import org.jetbrains.org.objectweb.asm.tree.AbstractInsnNode
 import org.jetbrains.org.objectweb.asm.tree.JumpInsnNode
 import org.jetbrains.org.objectweb.asm.tree.LabelNode
+import org.jetbrains.org.objectweb.asm.tree.LineNumberNode
 import org.jetbrains.org.objectweb.asm.tree.MethodNode
 
 class RedundantGotoMethodTransformer : MethodTransformer() {
     /**
-     * Removes redundant GOTO's, i.e. to subsequent labels
+     * Removes redundant GOTO's in the following cases:
+     *  (1) subsequent labels
+     *      ...
+     *      goto Label (can be removed)
+     *      Label:
+     *      ...
+     *  (2) indirect goto
+     *      ...
+     *      <branch instruction> Label (can be rewrote to <branch instruction> Label2)
+     *      ...
+     *      Label:
+     *      goto Label2 (must not be removed due to the previous instruction that can fallthrough on this goto)
+     *      ...
      */
     override fun transform(internalClassName: String, methodNode: MethodNode) {
         val insns = methodNode.instructions.toArray().apply { reverse() }
-        val insnsToRemove = arrayListOf<AbstractInsnNode>()
-
+        var insnsToRemove = arrayListOf<AbstractInsnNode>()
         val currentLabels = hashSetOf<LabelNode>()
+        var labelsToReplace = hashMapOf<LabelNode, JumpInsnNode>()
+        var pendingGoto: JumpInsnNode? = null
+
         for (insn in insns) {
             when {
-                insn is LabelNode ->
+                insn is LabelNode -> {
                     currentLabels.add(insn)
-                insn.opcode == Opcodes.GOTO &&
-                        (insn as JumpInsnNode).label in currentLabels ->
-                    insnsToRemove.add(insn)
-                insn.isMeaningful ->
+                    pendingGoto?.let { labelsToReplace[insn] = it }
+                }
+                insn.opcode == Opcodes.GOTO -> {
+                    pendingGoto = insn as JumpInsnNode
+                    if (insn.label in currentLabels) {
+                        insnsToRemove.add(insn)
+                    } else {
+                        currentLabels.clear()
+                    }
+                }
+                insn is LineNumberNode -> pendingGoto = null
+                insn.isMeaningful -> {
                     currentLabels.clear()
+                    pendingGoto = null
+                }
             }
         }
 
+        // Rewrite branch instructions.
+        if (!labelsToReplace.isEmpty()) {
+            insns.filter { it is JumpInsnNode }.forEach { rewriteLabelIfNeeded(it as JumpInsnNode, labelsToReplace) }
+        }
+
         for (insnToRemove in insnsToRemove) {
             methodNode.instructions.remove(insnToRemove)
         }
     }
+
+    private fun rewriteLabelIfNeeded(
+        jumpInsn: JumpInsnNode,
+        labelsToReplace: HashMap<LabelNode, JumpInsnNode>
+    ) {
+        getLastTargetJumpInsn(jumpInsn, labelsToReplace).takeIf { it.label != jumpInsn.label }?.let {
```

Comment:
Minor: I'd rather just use `if`. I find `takeIf` + `?.let` a bit more convoluted way to say basically the same thing: do something with a value if condition is met.

Reply:
Done